metadata-io/src/main/java/com/linkedin/metadata/entity/EntityServiceImpl.java

```diff
@@ -527,227 +481,192 @@ public ListResult<RecordTemplate> listLatestAspects(
+                } else {
+                  // patch to upsert
+                  PatchBatchItem patchBatchItem = (PatchBatchItem) item;
+                  final String urnStr = patchBatchItem.getUrn().toString();
+                  final EntityAspect latest = latestAspects.getOrDefault(urnStr, Map.of()).get(patchBatchItem.getAspectName());
+                  final RecordTemplate currentValue = latest != null
+                          ? EntityUtils.toAspectRecord(patchBatchItem.getUrn(), patchBatchItem.getAspectName(), latest.getMetadata(), _entityRegistry) : null;
+                  return patchBatchItem.applyPatch(_entityRegistry, currentValue);
+                }
+              })
+              .collect(Collectors.toList());
+
+      // Database Upsert results
+      List<UpdateAspectResult> upsertResults = items.stream()
+              .map(item -> {
+                final String urnStr = item.getUrn().toString();
+                final EntityAspect latest = latestAspects.getOrDefault(urnStr, Map.of()).get(item.getAspectName());
+                final long nextVersion = nextVersions.getOrDefault(urnStr, Map.of()).getOrDefault(item.getAspectName(), 0L);
+
+                final UpdateAspectResult result;
+                if (overwrite || latest == null) {
+                  result = ingestAspectToLocalDBNoTransaction(item.getUrn(), item.getAspectName(), item.getAspect(),
+                          auditStamp, item.getSystemMetadata(), latest, nextVersion).toBuilder().request(item).build();
+
+                  // support inner-batch upserts
+                  latestAspects.computeIfAbsent(urnStr, key -> new HashMap<>()).put(item.getAspectName(), item.toLatestEntityAspect(auditStamp));
+                  nextVersions.computeIfAbsent(urnStr, key -> new HashMap<>()).put(item.getAspectName(), nextVersion + 1);
+                } else {
+                  RecordTemplate oldValue = EntityUtils.toAspectRecord(item.getUrn().getEntityType(), item.getAspectName(),
+                          latest.getMetadata(), getEntityRegistry());
+                  SystemMetadata oldMetadata = EntityUtils.parseSystemMetadata(latest.getSystemMetadata());
+                  result = UpdateAspectResult.builder()
+                          .urn(item.getUrn())
+                          .request(item)
+                          .oldValue(oldValue)
+                          .newValue(oldValue)
+                          .oldSystemMetadata(oldMetadata)
+                          .newSystemMetadata(oldMetadata)
+                          .operation(MetadataAuditOperation.UPDATE)
+                          .auditStamp(auditStamp)
+                          .maxVersion(latest.getVersion())
+                          .build();
+                }
+
+                return result;
+              }).collect(Collectors.toList());
+
+      // commit upserts prior to retention or kafka send, if supported by impl
+      if (tx != null) {
+        tx.commitAndContinue();
```

Comment:
After the aspect insert, we commit before proceeding to the possible application of retention policy. If the retention logic fails, at least we still get the upserts of the aspects.